### ISSUE TITLE
[question] Should export_conandata_patches actually exports only a subset of patches instead of everything?

### ISSUE THREAD

**Issue by IgorIvanov-Kudan** [conan-io/conan#13874]
### What is your question?

I have been working on a tool to ensure that packages in our CCI fork are being built mostly with same Recipe Revisions, and noticed that some packages were getting different RREVs for no reason. Upon closer examinaton, I realized that problematic packages have in their `conanfile.py` this:
```python
    def export_sources(self):
        export_conandata_patches(self)
```
instead of
```python
  exports_sources = [ "patches/**" ]
```
, and something like this in `conandata.yml`:
```yaml
  "1.2.13":
    - patch_file: "patches/1.2.13/0001-Fix-cmake.patch"
    - patch_file: "patches/0002-gzguts-xcode12-compile-fix.patch"
  "1.2.11":
    - patch_file: "patches/1.2.11/0001-fix-cmake.patch"
    - patch_file: "patches/0002-gzguts-xcode12-compile-fix.patch"
    - patch_file: "patches/0003-gzguts-fix-widechar-condition.patch"
```
Finally, [`export_conandata_patches`](https://github.com/conan-io/conan/blob/release/1.59/conan/tools/files/patches.py#L125) added in #11911 copies only patch files corresponding to the version being currently built:
```python
    patches = conanfile.conan_data.get('patches')
    entries = patches.get(conanfile.version, [])
    for it in entries:
        patch_file = it.get("patch_file")
        if patch_file:
            src = os.path.join(conanfile.recipe_folder, patch_file)
            dst = os.path.join(conanfile.export_sources_folder, patch_file)
            shutil.copy2(src, dst)
```
---
Since Conan uses patches to calculate RREV, and since a different set of patches will be exported for versions `1.2.11` and `1.2.13`, I suddenly got a report from my tool that our `zlib` is uploaded under two different RREVs:
```
INFO  :: processing JSON outputs from `conan create` into YML metadata file...
INFO  ::   flattening `conan create` outputs...
INFO  ::     #1 - zlib/1.2.13@KdConanRecipes/pr-219#d04fdad4cc914aeaed6c7c3df56a35e6
...
INFO  ::     #24 - zlib/1.2.11@KdConanRecipes/pr-219#c721ca6f26833f786fc76e196c6158d5
INFO  ::   checking for conflicting RREVs...
ERROR ::     All packages should have same reference and RREV! Conflicts found:
- zlib/1.2.11#c721ca6f26833f786fc76e196c6158d5 <-> zlib/1.2.13#d04fdad4cc914aeaed6c7c3df56a35e6
```

I have looked into #11911, and it seems that the function (and corresponding changes in recipes) was added to solve a specific issue #11770:

> If we export patches with `exports_sources = "patches/**"`, each version ends up with patches of unrelated versions, and adding or modifying a patch of a version can change RREV of other versions.

Since @SpaceIm has much greater experience in maintaining CCI than me, I assume I am doing something wrong. But as a maintainer I would have actually expected RREVs to instead be same for different package versions. At least as long as they are built from the same recipe and nothing changes drastically between versions.

Since usually a single recipe is used to build a multitude of versions, it always made sense to me to see all patches exported. I kinda assume it to be just a handful "extension" of `conanfile.py` - instead of having `if version == 1.0: tool.replace(file, 'some super long multiline foo', 'some shorter singleline bar')` in Python we just form it as an actual patch file and apply it using existing tools for patch application.

I expected to have a small number of exceptions in my tool for packages like `openssl`, where you have 2 separate recipes to build versions `1.x` and `3.x`, and maybe for something else where a different set of actual sources is being exported between versions. Now I am not sure whether I should just unconditionally allow RREVs to differ between versions, or rollback a bunch of recipes to use `export_sources = ["patches/**"]` instead of `export_conandata_patches()`

---

The main point of my rant is: I would assume it to also cause issues with resolving correct versions of binaries - when the remote has `RREV=d04f` marked as most recent with binaries for `1.2.13`, and requesting `1.2.11` would fail because corresponding binaries were uploaded for "outdated" `RREV=c721`. But either Conan manages to handle that, or neither we nor other users have yet stumbled upon such issue.

I guess I would just like to get some confirmation, that despite my concerns the behavior "export with version X only patches for version X" is indeed intended, and does not cause any longterm issues with RREV mismatches, even if build process does not yet involve Lockfiles. Because for us it still does not - I am actually upgrading our package recipes first, in order to switch to Conan 2.0, in order to finally achieve usable Lockfiles.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

---
*(oops, wrong account!* :facepalm: *it's a me, @Artalus)*

**Comment by memsharded** [comment]
Hi @IgorIvanov-Kudan 

Thanks for the report.
I think the main issue here would be that such a layout and structure relies on using the ConanCenter hooks, specially the one that is trimming/removing the bits from ``conandata.yml`` not necessary for the current version, so this is not a built-in feature in the recipes, but rely on the hook.

I agree that this is not optimal, and I had planned thinking about it with the team to see what could be possible improvements.

**Comment by memsharded** [comment]
Hi @IgorIvanov-Kudan 

I have rechecked above comments, and I want to clarify:

- ``export_conandata_patches()`` exports only the subset of patches instead of everything. This is already the current behavior.
- This behavior is critical for stability. Otherwise, adding a new version to the ``conandata.yml`` will break all previous published versions, inducing new recipe revisions and requiring re-building all binaries for all the previous versions. This is extremely undesired and considered a bug by all users: a new version in ``conandata.yml`` should not require re-building from source all previous versions of that same package. This would be unaffordable in ConanCenter CI (imagine having to re-build all previous versions of boost, for all configurations, every time a new version of boost is added. Or for opencv or qt). But it will also break other things because of creating new artificial recipe revisions, that could also imply rebuilding many consumer packages downstream.
- Different versions of the same library will always have different revisions. It is impossible not have that behavior. Because there must be something in the recipe that is different for the different versions. In the ConanCenter case, the ``zlib/1.2.11`` will contain a final ``conandata.yml`` that will contain something like ``https://zlib.net/..../zlib1.2.11.tgz``, while the ``conandata.yml`` for ``zlib/1.2.13`` will contain something like ``https://zlib/.../zlib/1.2.13.tgz``. Having different ``conandata.yml`` will expectedly induce different revisions.

In summary, having different revisions for different versions of the same library is not only expected, but totally necessary and fundamental to Conan behavior.


What I thought was the issue was that the hook that trims conandata.yml was not activated, but this was not the issue. Labeling this as question.

**Comment by IgorIvanov-Kudan** [comment]
> adding a new version to the conandata.yml will break all previous published versions

Can you elaborate, why does that happen? I see only that it will cause a revision mismatch so "older recipe" won't have binaries for new version, and vice versa "newer recipe" won't have binaries for old versions - but that would be essentially the same as having each recipe to have only a specific part of conandata.

> Different versions of the same library will always have different revisions. It is impossible not have that behavior. Because there must be something in the recipe that is different for the different versions.

You mean globally, "in principle", or as a result of the changes targeting the conandata behavior? Because recipes used to just export everything "statically" via `exports = [...]` and `exports_sources = [...]`, so both `zlib/1.2.11` and `zlib/1.2.13` would have the same recipe, conandata, patches and everything.

**Comment by memsharded** [comment]
> Can you elaborate, why does that happen? I see only that it will cause a revision mismatch so "older recipe" won't have binaries for new version, and vice versa "newer recipe" won't have binaries for old versions - but that would be essentially the same as having each recipe to have only a specific part of conandata.

Because adding a new version, if not removed from ``conandata.yml`` and the patches are filtered, creates a new revision for every previously existing version. Further analysis of what did change is not that straightforward, so it must be assumed that the recipe got some changes, and it is a new revision, so it must be built from sources. That means:
- Tons of resources are wasted in ConanCenter, rebuilding from source all previous versions of every package when a new version is added
- The creating of new recipe revisions for all those previous versions will also break every user that depends on them with ``full_package_mode`` or any mode that includes the recipe revision.

> You mean globally, "in principle", or as a result of the changes targeting the conandata behavior? Because recipes used to just export everything "statically" via exports = [...] and exports_sources = [...], so both zlib/1.2.11 and zlib/1.2.13 would have the same recipe, conandata, patches and everything.

This is impossible. 2 versions cannot have exactly the same sources, exports, exports sources and recipe, otherwise they would be building exactly the same version of the library. There must be something there that is different at least 1.2.11<->1.2.13 difference will happen in the exported recipe, and that will inevitably cause a different recipe revision. The only way would be that they always export all the conandata.yml and patches, for every version, but as described above, that breaks the assumption that adding a new version to conandata.yml shouldn't require re-building all previous versions from sources.

**Comment by IgorIvanov-Kudan** [comment]
> - The creating of new recipe revisions for all those previous versions will also break every user that depends on them with `full_package_mode` or any mode that includes the recipe revision.

Ah... I see now, yes, these would definitely break.

So from the point of view of a CCI fork maintainer, if we
1. do not use `full_package_mode` anywhere in our enterprise environment,
1. do not support a wide range of versions / can afford to rebuild everything mentioned in `recipes/<pkg>/config.yml` on each recipe update,
1. and remove older versions from `config.yml` as we add newer ones,

— everything should be alright on our side? But if we cannot guarantee `#1` or afford `#2`, we should activate [`hook_reduce_conandata.py`](https://github.com/conan-io/hooks/blob/master/hooks/hook_reduce_conandata.py) in our package CI, or at least implement some other mechanism for checking what was changed in recipe files?

I guess my main concern is keeping everything (including older commits, to some sane extent) buildable, while also wanting to maintain some transparency in Artifactory.
If for example we drop `zlib/1.2.13` from config and add `zlib/1.2.14`. Would Conan still be able to use both `1.2.14` for newer commits and `1.2.13` for older commits? Despite them having different RREVs, whether we strip conandata or no.

**Comment by memsharded** [comment]
> I guess my main concern is keeping everything (including older commits, to some sane extent) buildable, while also wanting to maintain some transparency in Artifactory.

If you drop ``zlib/1.2.13`` from ``config.yml``, but you leave it in ``conandata.yml``, but you don't apply  the hook to reduce ``conandata.yml`` you will also have things not in sync, because what you would get from ``conan create --version=1.2.13`` in the repo would be a different recipe revision than what you have in Artifactory, because you didn't build it, you just dropped from ``config.yml`` and didn't rebuild it. If you remove the ```1.2.13`` from ``conandata.yml`` you just break the creation of the package from source, but the package binaries are still in Artifactory, unless you explicitly remove it.

> If for example we drop zlib/1.2.13 from config and add zlib/1.2.14. Would Conan still be able to use both 1.2.14 for newer commits and 1.2.13 for older commits? Despite them having different RREVs, whether we strip conandata or no.

I am afraid I don't understand this point. Conan will use 1.2.14 if there is a ``requires = "zlib/1.2.14"``, and it will use the older one if there is a ``requires = "zlib/1.2.13"``. Not sure how this is connected to revisions.

**Comment by IgorIvanov-Kudan** [comment]
> Not sure how this is connected to revisions.

Indeed, sorry! I actually forgot that the packages are grouped in remote by folders in order of `package / version / rrev`, and was instead worrying about `package / rrev / version` hierarchy for some reason.

Thank you for your time, you've clarified all the questions I had and then some :smile:
